# Laboratorio 2 - Reglas de Asociación y Patrones Secuenciales

## Grupo 1 - Paula Hípola Gómez, José Ignacio Navas Sanz y Belén Ortega Pérez

## Entregable 1 - Reglas de Asociación

### 1.1 (1 punto) Estudia brevemente el dataset (número de usuarios, número de atributos, categorías más frecuentes para cada atributo/columna, etc.) y realiza el preprocesamiento para generar el dataset necesario para aplicar el algoritmo Apriori a este conjunto de datos.

In [2]:
%config IPCompleter.greedy=True
%reset
#Importamos las librerías necesarias
#Numpy
import numpy as np 
#Pandas
import pandas as pd
#Scikit-Learn
import sklearn as sk
#Matplotlib
import matplotlib.pyplot as plt

Once deleted, variables cannot be recovered. Proceed (y/[n])? y


### Cargamos los datos y los visualizamos

In [3]:
#Importamos el CSV
dataset = pd.read_csv('radio.csv')

#Mostramos los datos
dataset.head()

user                   artist sex  country
0     1    red hot chili peppers   f  Germany
1     1  the black dahlia murder   f  Germany
2     1                goldfrapp   f  Germany
3     1         dropkick murphys   f  Germany
4     1                 le tigre   f  Germany

In [4]:
print('La dimensión del dataset es:',dataset.shape)

La dimensión del dataset es: (289955, 4)


Tenemos 289955 instancias y 4 atributos, para cada una de nuestras instancias. Analizando más profundamente el dataset, está compuesto con las interacciones de cada usuario que ha escuchado al menos a un artista. Vamos a ver cuantos valores tenemos actualmente en cuanto a usuarios, artistas y demás atributos y luego procederemos a identificar en el dataset valores duplicados y nulos que dificultan la comprensión de los datos y procederemos a su eliminación.

In [5]:
numUsuarios = dataset.groupby('user').count().shape[0]
print('Hay',numUsuarios,'usuarios.')
numArtistas = dataset.groupby('artist').count().shape[0]
print('Hay',numArtistas,'artistas.')
numSexos = dataset.groupby('sex').count().shape[0]
print('Hay',numSexos,'géneros.')
numPaises = dataset.groupby('country').count().shape[0]
print('Hay',numPaises,'países.')

Hay 15000 usuarios.
Hay 1004 artistas.
Hay 2 géneros.
Hay 159 países.


Como podemos ver, hemos agrupado todos los usuarios, artistas, géneros y países en sus respectivas categorías para comprender mejor el dataset planteado, ahora procederemos a preprocesar y limpiar los datos.

### Limpieza de datos

In [6]:
dataset.duplicated(keep=False)

0         False
1         False
2         False
3         False
4         False
          ...  
289950    False
289951    False
289952    False
289953    False
289954    False
Length: 289955, dtype: bool

In [7]:
#Eliminación de nulos
dataset = dataset.dropna()
print('Tamaño sin nulos',dataset.shape)

#Eliminación de duplicados
dataset = dataset.drop_duplicates()
print('Tamaño sin duplicados',dataset.shape)
dataset

Tamaño sin nulos (289955, 4)
Tamaño sin duplicados (289953, 4)


user                   artist sex  country
0           1    red hot chili peppers   f  Germany
1           1  the black dahlia murder   f  Germany
2           1                goldfrapp   f  Germany
3           1         dropkick murphys   f  Germany
4           1                 le tigre   f  Germany
...       ...                      ...  ..      ...
289950  19718                bob dylan   f   Canada
289951  19718                   pixies   f   Canada
289952  19718                the clash   f   Canada
289953  19718     a tribe called quest   f   Canada
289954  19718                radiohead   f   Canada

[289953 rows x 4 columns]

Tenemos que hacer un nuevo dataset para analizar estos datos con el algoritmo Apriori, para ello vamos a construir un dataset con un listado de ususarios con todos los artistas que ha escuchado...

In [12]:
dagrupado = dataset.groupby('user')['artist'].apply(list)
dagrupado

user
1        [red hot chili peppers, the black dahlia murde...
3        [devendra banhart, boards of canada, cocorosie...
4        [tv on the radio, tool, kyuss, dj shadow, air,...
5        [dream theater, ac/dc, metallica, iron maiden,...
6        [lily allen, kanye west, sigur rós, pink floyd...
                               ...                        
19713    [armin van buuren, above & beyond, atb, ferry ...
19714    [misfits, type o negative, arch enemy, red hot...
19715    [abba, james blunt, jason mraz, amy winehouse,...
19717    [marilyn manson, beyoncé, madonna, t.a.t.u., k...
19718    [beirut, of montreal, black flag, the new porn...
Name: artist, Length: 15000, dtype: object

### 1.2 (4 puntos) Aprende a manejar la librería mlextend y su implementación del algoritmo Apriori.
#### 1.2.1 Obtén los itemsets frecuentes para k=1. Para esto es necesario obtener el soporte de los itemset. Crea una función que utilice métodos de la librería mlextend de manera que dado un itemset devuelva su soporte.



In [13]:
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(dagrupado).transform(dagrupado)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

...and you will know us by the trail of dead   2pac  3 doors down  \
0                                             False  False         False   
1                                             False  False         False   
2                                             False  False         False   
3                                             False  False         False   
4                                             False  False         False   
...                                             ...    ...           ...   
14995                                         False  False         False   
14996                                         False  False         False   
14997                                         False  False         False   
14998                                         False  False         False   
14999                                         False  False         False   

       30 seconds to mars    311  36 crazyfists     44  50 cent  \
0                   False  False          False  False    False   
1                   False  False          False  False    False   
2                   False  False          False  False    False   
3                   False  False          False  False    False   
4                   False  False          False  False    False   
...                   ...    ...            ...    ...      ...   
14995               False  False          False  False    False   
14996               False  False          False  False    False   
14997               False  False          False  False    False   
14998               False  False          False  False    False   
14999               False  False          False  False    False   

       65daysofstatic  [unknown]  ...  wilco  within temptation  \
0               False      False  ...  False              False   
1               False      False  ...  False              False   
2               False      False  ...  False              False   
3               False      False  ...  False              False   
4               False      False  ...  False              False   
...               ...        ...  ...    ...                ...   
14995           False      False  ...  False              False   
14996           False      False  ...  False              False   
14997           False      False  ...  False              False   
14998           False      False  ...  False              False   
14999           False      False  ...  False              False   

       wolfgang amadeus mozart  wu-tang clan  yann tiersen  yeah yeah yeahs  \
0                        False         False         False            False   
1                        False         False         False            False   
2                        False         False         False            False   
3                        False         False         False            False   
4                        False         False         False            False   
...                        ...           ...           ...              ...   
14995                    False         False         False            False   
14996                    False         False         False            False   
14997                     True         False         False            False   
14998                    False         False         False            False   
14999                    False         False         False             True   

       yellowcard  yo la tengo  zero 7  Édith piaf  
0           False        False   False       False  
1           False        False   False       False  
2           False        False   False       False  
3           False        False   False       False  
4           False        False   False       False  
...           ...          ...     ...         ...  
14995       False        False   False       False  
14996       False        False   False       False  
14997       False        False   False       False  
14998

#### 1.2.2. Crea una función que devuelva los itemsets frecuentes candidatos y su soporte para k>=2



#### 1.2.3. Crea una función que repita el paso 1.2.2. hasta que no se generen nuevos itemsets frecuentes



#### 1.2.4. Crea una función que muestre todas las reglas posibles con su confianza



#### 1.2.5. Lista todas las reglas que sean de alta confianza (p.ej. con valor >=0.5)



#### 1.2.6. Crea funciones que permitan lo siguiente:
<ul>
    <li><b>* devolver todas las reglas que contengan un antecesor dado</b></li>
    <li><b>* devolver todas las reglas que tengan una confianza mayor (o igual) que un umbral mínimo dado</b></li>
    <li><b>* devolver todas las reglas que tengan un lift en una de las tres franjas de estudio (under 1, over 1, =1)</b></li>
</ul>



#### 1.2.7. Utiliza al menos dos representaciones gráficas para mostrar las reglas obtenidas e interpretar los datos